src/utils/ScheduledQueue.hxx

```diff
@@ -0,0 +1,218 @@
+    /// @param strides is an array of the size num_bands. It contains the
+    /// stride coefficients (they should all be numbers <= 1. A value of 1
+    /// degrades to strict priority order). This array is not used anymore
+    /// after the constructor returns. A stride of 0.2 means that 20% of all
+    /// tokens at this level will be allocated to this band, and 80% of the
+    /// tokens are passed down to lower priorities (bands with higher index).
+    ScheduledQueue(unsigned num_bands, const Fixed16 *strides)
+        : numBands_(num_bands)
+        , bands_(new Band[num_bands])
+    {
+        for (unsigned i = 0; i < numBands_; i++)
+        {
+            bands_[i].stride_ = strides[i];
+            bands_[i].currentToken_ -= strides[i];
+        }
+    }
+
+    /// Destructor.
+    ~ScheduledQueue()
+    {
+        delete[] bands_;
+    }
+
+    /// Get an item from the queue. The returned item will be according to the
+    /// priority scheduler.
+    /// @return the member and the priority from which it came.
+    Result next()
+    {
+        OSMutexLock h(lock());
+        return next_locked();
+    }
+
+    /// Get an item from the queue. The returned item will be according to the
+    /// priority scheduler. The caller must acquire the lock first.
+    /// @return the member and the priority from which it came.
+    Result next_locked()
+    {
+        if (!numPending_)
+        {
+            // Empty queue.
+            return Result(0, 0);
+        }
+        // Execute the priority based scheduling algorithm.
+        for (unsigned i = 0; i < numBands_; ++i)
+        {
+            bands_[i].currentToken_ += bands_[i].stride_;
+            if (bands_[i].currentToken_.trunc() >= 1)
+            {
+                Result ret = bands_[i].queue_.next_locked();
+                if (ret.item)
+                {
+                    ret.index = i;
+                    --numPending_;
+                    bands_[i].currentToken_ -= 1;
+                    return ret;
+                }
+                else
+                {
+                    // This queue has a token but is empty. We remove
+                    // fractional tokens and keep searching onwards in the
+                    // priorities.
+                    bands_[i].currentToken_ = 1;
+                }
+            }
+        }
+        // Fallen off at the end. We go backwards to find any queue with
+        // nonempty members.
+        for (int i = numBands_ - 1; i >= 0; --i)
+        {
+            if (!bands_[i].queue_.empty())
+            {
+                Result ret = bands_[i].queue_.next_locked();
+                bands_[i].currentToken_ = 0;
+                ret.index = i;
+                --numPending_;
+                return ret;
+            }
+        }
+        DIE("Unexpected nonempty queue");
+        return Result(0, 0);
+    }
+
+    /// @return the lock to use for the _locked() functions.
```

Comment:
Why is this public? Is the user supposed to use this lock or is it only used inside of object members? If the user is expected to use this lock, can you add some comments to indicate where and why?

Reply:
Added more verbose comments.
```
    /// The caller must acquire this lock before using any of the _locked()
    /// functions. If the caller needs to do many operations in quick
    /// succession, it might be faster to do them under a single lock,
    /// i.e. acquire lock() first, then call xxx_locked() repeatedly, then
    /// unlock.
    /// @return the lock to use for the _locked() functions.
```

This API also exists on Queue classes (it's also public there), and I wanted to keep a compatible API.